reflekt-plugin/src/main/kotlin/org/jetbrains/reflekt/plugin/analysis/collector/ir/InstancesCollector.kt

```diff
@@ -35,9 +41,49 @@ class InstancesCollector(
  */
 class InstancesCollectorExtension(
     private val irInstancesAnalyzer: IrInstancesAnalyzer,
+    private val libraryArgumentsWithInstances: LibraryArgumentsWithInstances,
+    private val reflektMetaFilesFromLibraries: Set<File>,
     private val messageCollector: MessageCollector? = null,
 ) : IrGenerationExtension {
+    // TODO: can we avoid making a copy here?
+    private var libraryArguments = LibraryArguments()
+    private var irInstancesFqNames = IrInstancesFqNames()
+
     override fun generate(moduleFragment: IrModuleFragment, pluginContext: IrPluginContext) {
+        extractInstancesFromLibraries(pluginContext)
         moduleFragment.acceptChildrenVoid(InstancesCollector(irInstancesAnalyzer, messageCollector))
     }
+
+    private fun extractInstancesFromLibraries(pluginContext: IrPluginContext) {
+        reflektMetaFilesFromLibraries.forEach { metaFile ->
+            val currentLibraryArgumentsWithInstances = SerializationUtils.decodeArguments(metaFile.readBytes(), pluginContext)
+            libraryArguments = libraryArguments.merge(currentLibraryArgumentsWithInstances.libraryArguments)
+            irInstancesFqNames = irInstancesFqNames.merge(currentLibraryArgumentsWithInstances.instances)
+        }
+        libraryArgumentsWithInstances.replace(libraryArguments, irInstancesFqNames)
+    }
+}
+
+class LibraryInstancesCollectorExtension(
+    private val irInstancesAnalyzer: IrInstancesAnalyzer,
+    private val irInstancesFqNames: IrInstancesFqNames,
+) : IrGenerationExtension {
+    private val externalLibraryId = "EXTERNAL_LIBRARY"
+
+    override fun generate(moduleFragment: IrModuleFragment, pluginContext: IrPluginContext) {
+        val classes = findIrClasses(irInstancesFqNames.classes, pluginContext::referenceClass)
+        val objects = findIrClasses(irInstancesFqNames.objects, pluginContext::referenceClass)
+        val functions = findIrFunctions(irInstancesFqNames.functions, pluginContext::referenceFunctions)
+        irInstancesAnalyzer.addDeclarations(externalLibraryId, classes, objects, functions)
```

Comment:
We can not extract the file id from the library. And for the libraries, it is ok since we store all entities by files for supporting incremental compilation in the future.